### PR TITLE
Friends App

### PR DESCRIPTION
**Pull request by YuliiaHordiichuk**
### Friends App

[Demo](https://yuliiahordiichuk.github.io/Friends-app/)
[Code base](https://github.com/YuliiaHordiichuk/Friends-app)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+                }
+                if (a.name.first < b.name.first) {
+                    return 1; 
+                }
+                return 0; 
+            })
+        }
+
+    return dataArray; 
+}
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:104]: You should split this huge function into several separate - by age, by name, etc.

**Review by zonzujiro — CHANGES_REQUESTED** [review]
I see some common mistakes - pls check this AMA and fix code using information from it https://www.youtube.com/watch?v=iqdUtYGYdDE

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 2]
```diff
@@ -0,0 +1,130 @@
+const randomUsersUrl = 'https://randomuser.me/api/?results=36&inc=gender,name,email,phone,picture,dob';
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:1]: Why it is global?

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        })
+}; 
+
+function initApp() {
+    getData(); 
+    bindEventListeners(); 
+}
+
+function bindEventListeners () {
+    const form = document.querySelector('.form');
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:27]: Please, give less generic class name to form

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+        newUsers = sortByName(newUsers)
+    }
+
+    return newUsers; 
+}
+
+function createCards(dataArray) {
+    const cardsContainer = document.querySelector('.cards__items'); 
+
+    cardsContainer.innerHTML = dataArray.map(( { picture: {large}, name: { first, last }, dob: {age}, email, phone } ) => {
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:66]: Prettier? 🤔

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    } else if (sortBy === '9-0') {
+        sortByOrder(dataArray, 'dob', 'age', 'descending')
+    }
+
+    return dataArray;
+}
+
+function sortByName(dataArray) {
+    if (sortBy === 'a-z') {
+        sortByOrder (dataArray, 'name', 'first', 'ascending')
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:92]: `'name', 'first', 'ascending'` - ok, this part bad. These are called magic strings - unnamed string variables that are used as arguments.

First of all, usually, in JS, you should not use strings as arguments.
Second, an idea where you pass keys for an object which should be filtered or sorted also bad. What if the object will be changed? What if you will misprint in keys?

Pls, check other closed PRs and find better way to sort your data

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+function sortByOrder(dataArray, keyToCompare, valueToCompare, sortType) {
+    dataArray.sort((a,b) => {
+        if (a[keyToCompare][valueToCompare] < b[keyToCompare][valueToCompare]) {
+            return -1; 
+        }
+        if (a[keyToCompare][valueToCompare] > b[keyToCompare][valueToCompare]) {
+            return 1; 
+        }
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:107]: As I told, nope :)

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+        return 0; 
+    })
+
+    if (sortType === 'ascending') {
+        return dataArray; 
+    } else if (sortType === 'descending') {
+        return dataArray.reverse(); 
+    }
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:115]:
```suggestion
  return sortType === 'descending' ? dataArray.reverse() : dataArray;
```

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+    if (sortBy === '0-9') {
+        sortByOrder(dataArray, 'dob', 'age', 'ascending')
+    } else if (sortBy === '9-0') {
+        sortByOrder(dataArray, 'dob', 'age', 'descending')
+    }
+
+    return dataArray;
+}
+
+function sortByName(dataArray) {
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:90]: To generic name for variable

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 9]
```diff
@@ -0,0 +1,130 @@
+const randomUsersUrl = 'https://randomuser.me/api/?results=36&inc=gender,name,email,phone,picture,dob';  
+
+let users; 
+let searchBy = ''; 
+let sortBy = ''; 
+let filterBy = 'all'; 
+ 
+function getData() {
+    fetch(randomUsersUrl)
+        .then(answer => answer.json())
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:10]: https://www.tjvantoll.com/2015/09/13/fetch-and-errors/

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+        .then(( { results } ) => {
+            users = results;
+            createCards(results); 
+        })
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:14]: This should be done outside of the function which gets data.

Pls, read about single responsibility. Right now function named `getData` will also create cards.

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+                <p class="card__detail">${age} years old </p>
+                <h3 class="card__subtitle">Contact details</h3>
+                <a class="card__detail card__detail_link" href="mailto:${email}">${email}</a>
+                <a class="card__detail card__detail_link" href="tel:${phone}">${phone}</a>
+            </li>
+        `; 
+    }).join(''); 
+}
+
+function sortByAge(dataArray) {
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:80]: Generic name

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 12]
```diff
@@ -0,0 +1,130 @@
+const randomUsersUrl = 'https://randomuser.me/api/?results=36&inc=gender,name,email,phone,picture,dob';  
+
+let users; 
+let searchBy = ''; 
+let sortBy = ''; 
+let filterBy = 'all'; 
+ 
+function getData() {
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:8]: Generic name

File: submissions/YuliiaHordiichuk/friends-app/app.js [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+    if (sortBy === 'a-z') {
+        sortByOrder (dataArray, 'name', 'first', 'ascending')
+    } else if (sortBy === 'z-a') {
+        sortByOrder (dataArray, 'name', 'first', 'descending')
+    }
+
+    return dataArray; 
+}
+
+function sortByOrder(dataArray, keyToCompare, valueToCompare, sortType) {
```
Comment (zonzujiro) [inline submissions/YuliiaHordiichuk/friends-app/app.js:100]: I have a strong filling that you just copy-pasted this code from one of the works of our students 🤔

**Review by zonzujiro — CHANGES_REQUESTED** [review]

File: submissions/YuliiaHordiichuk/friends-app/app.js (hunk 2 above)
Comment (YuliiaHordiichuk) [reply submissions/YuliiaHordiichuk/friends-app/app.js:1]: Moved

File: submissions/YuliiaHordiichuk/friends-app/app.js (hunk 9 above)
Comment (YuliiaHordiichuk) [reply submissions/YuliiaHordiichuk/friends-app/app.js:10]: Thanks, very helpful!

File: submissions/YuliiaHordiichuk/friends-app/app.js (hunk 5 above)
Comment (YuliiaHordiichuk) [reply submissions/YuliiaHordiichuk/friends-app/app.js:92]: Got it, thanks. I  changed the  functions which return sorted users.

File: submissions/YuliiaHordiichuk/friends-app/app.js (hunk 13 above)
Comment (YuliiaHordiichuk) [reply submissions/YuliiaHordiichuk/friends-app/app.js:100]: I've seen similar approaches to sorting data, but I haven't copied this code.

File: submissions/YuliiaHordiichuk/friends-app/app.js (hunk 6 above)
Comment (YuliiaHordiichuk) [reply submissions/YuliiaHordiichuk/friends-app/app.js:107]: Agreed :)

**Pull request merged**